megatron/core/extensions/gemm_patch.py

```diff
@@ -0,0 +1,182 @@
+    MXFP8TensorStorage,
+)
+from transformer_engine.pytorch.tensor.storage.nvfp4_tensor_storage import (  # type: ignore
+    NVFP4TensorStorage,
+)
+
+from megatron.core.utils import log_single_rank
+
+logger = logging.getLogger(__name__)
+
+_GLOBAL_IS_PATCHED_ = None
+
+orig_general_gemm = transformer_engine.pytorch.cpp_extensions.gemm.general_gemm
+orig_general_grouped_gemm = transformer_engine.pytorch.cpp_extensions.gemm.general_grouped_gemm
+
+CKSUM_TOLERANCE = 0.01
+EPSILON = 1e-5
+
+DTYPES = {torch.bfloat16: "bf16", torch.float16: "fp16", torch.float32: "fp32"}
+
+
+def get_rank_info():
+    """Return rank and device info for logging."""
+    rank = os.environ.get("RANK", 0)
+    node: str = os.uname()[1]
+    device: int = torch.cuda.current_device()
+    return f"{rank}|{node}|{device}"
+
+
+def get_dtype(t):
+    """Return dtype info about a quantized tensor."""
+    if isinstance(t, (MXFP8Tensor, MXFP8TensorStorage)):
+        return "mxfp8"
+    elif isinstance(t, (NVFP4Tensor, NVFP4TensorStorage)):
+        return "nvfp4"
+    elif isinstance(t, torch.Tensor):
+        return DTYPES.get(t.dtype, "torch-unknown")
+    else:
+        return "unknown"
+
+
+def new_general_gemm(A, B, workspace, out=None, accumulate=False, bias=None, **kwargs):
+    """
+    A wrapper for general_gemm that calls the original method twice
+    and compares results for consistency.
+    """
+    cksum = [0, 0]
+    # Run twice for comparison.
+    for run_idx in range(2):
+        is_first_iter = run_idx == 0
+        need_output_copy = out is not None and (bias is out or accumulate)
+        if need_output_copy and is_first_iter:
+            out_arg = out.clone().detach()
+            if bias == out:
+                bias_arg = out_arg
+            else:
+                bias_arg = bias
+        else:
+            out_arg = out
+            bias_arg = bias
+
+        res = orig_general_gemm(
+            A, B, workspace, out=out_arg, accumulate=accumulate, bias=bias_arg, **kwargs
+        )
+        out_res = res[0]
+        cksum[run_idx] = out_res.sum(dtype=torch.float32).item()
+        if is_first_iter:
+            rerun_res = out_res
+            continue
+
+    # Check the consistency of the first and second run and return.
+    same_cksum = math.fabs(cksum[1] - cksum[0]) / (math.fabs(cksum[1]) + EPSILON) <= CKSUM_TOLERANCE
+    if not same_cksum:
+        logger.log(
+            logging.ERROR,
+            f"RANK {get_rank_info()}: DIFFERENT CHECKSUM ON 2ND RUN for dtype "
+            f"{get_dtype(A)}. {cksum[1]} != {cksum[0]}, "
+            f"shapes A:{list(A.size())},B:{list(B.size())},out:{list(out_res.shape)}",
+        )
+    elif not torch.equal(rerun_res, out_res):
+        logger.log(
+            logging.ERROR,
+            f"RANK {get_rank_info()}: DIFFERENT BITWISE on 2ND RUN for dtype "
+            f"{get_dtype(A)}. "
+            f"shapes A:{list(A.size())},B:{list(B.size())},out:{list(out_res.shape)}",
```

Comment:
```suggestion
            f"RANK {get_rank_info()}: DIFFERENT BITWISE on 2ND RUN for operation "
            f"A:({get_dtype(A)} , {list(A.size())}) x B:({get_dtype(B)} , {list(B.size())}) "
            f"-> out:({get_dtype(out_res)} , {list(out_res.size())})",
```
Logging this too as it's very useful